languages/julia/concepts/iterator-protocol/about.md

```diff
@@ -0,0 +1,92 @@
+<!-- TODO: Motivate why iterators are useful -->
+<!-- TODO: Add explanation why the Base.-prefix or import is necessary -->
+
+To implement the informal iteration interface, you need to define two methods `iterate(iter::MyIter)` and `iterate(iter::MyIter, state)`.
+Both methods must return a tuple of an item and the state of the iterator.
+The first method will return the first item and state, while the second method will return the next item and state.
+The iteration ends if `iterate` returns `nothing`.
+
+!!! note
+    Julia currently doesn't have a way to formally define interfaces.
+    To implement an interface, you need to look up which methods need to be defined in the Manual, or in the documentation of the package that "defines" the interface.
+    Alternatively, you can work your way through `MethodErrors` to find out which methods you need to implement:
+    <!-- TODO check if code blocks work within info boxes -->
+    ```julia
+    julia> struct MyIter end;
+
+    julia> for i in MyIter()
+            println(i)
+        end
+    ERROR: MethodError: no method matching iterate(::MyIter)
+    ...
+    ```
+
+!!! note
+    The iterator object itself is usually not mutated by iteration.
+
+To make it clearer when the `iterate` methods are called, one can look at the translation of the `for`-loop syntax:
+
+```julia
+for item in iter
+    println(item)
+end
+```
+
+will be translated to
+
+```julia
+next_item = iterate(iter)
+
+while !isnothing(next_item)
+    (item, state) = next_item
+    
+    println(item)
+
+    next_item = iterate(iter, state)
+end
+```
+
+## Example
```

Comment:
What's the appropriate level here, @ErikSchierboom or @iHiD?

Reply:
`##` for the top level headings is correct, so if "Example" is a top-level heading within this document (on par with "Advanced Usage"), then this is correct.

Reply:
It's meant to be the 2nd level. The top level heading would be "Iterator Protocol" which I understand is added outside of this file?

Reply:
Correct!